### ISSUE TITLE
Add client manager to access available client dynamically

### ISSUE THREAD

**Issue by ajstevens** [Drenso/symfony-oidc#13]
I am trying to have multiple firewalls in Symfony with each their own OIDC config. 
In the firewall I specify the client: 

```yaml

        rkd:
            pattern: ^/rkd
            lazy: true
            provider: oidc_user_provider
            oidc:
                client: rkd
                user_identifier_property: 'email'
                login_path: '/rkd/login_oidc'
                check_path: '/rkd/login_check'
            logout:
                path: logout

        admin:
            pattern: ^/admin
            lazy: true
            provider: oidc_user_provider
            oidc:
                client: admin
                user_identifier_property: 'email'
                login_path: '/admin/login_oidc'
                check_path: '/admin/login_check'
            logout:
                path: logout

```
I have multiple clients defined, which I can load with DI if i know the client name in the DI OIDCClientInterface the login url.  But I would like to get a specific named client dynamically. 
```
    default: 
      # Required OIDC client configuration
      well_known_url: '%env(OIDC_WELL_KNOWN_URL)%'
      client_id: '%env(OIDC_CLIENT_ID)%'
      client_secret: '%env(OIDC_CLIENT_SECRET)%'
      redirect_route: '/rkd/login_check'

    admin: 
      # Required OIDC client configuration
      well_known_url: '%env(OIDC_WELL_KNOWN_URL)%'
      client_id: 'cap_rkd'
      client_secret: '%env(OIDC_CLIENT_SECRET)%'
      redirect_route: '/test/login_check'
```

The problem is with starting the OIDC, I always get the default client in the login controller. 
```php
  #[Route('/{slug}/login_oidc', name: 'login_oidc')]
  #[IsGranted('PUBLIC_ACCESS')]
    public function surfconext(Organisation $organisation, OidcClientInterface $client): RedirectResponse
    {
        return $client->generateAuthorizationRedirect();   
    }
```

How can I best get the specific client in this case?

**Comment by bobvandevijver** [comment]
As per [doc](https://github.com/Drenso/symfony-oidc#oidc-clients), you can request non-default clients by using the correct parameter name.

> Clients are autowirable by using OidcClientInterface ${name}OidcClient, for example OidcClientInterface $defaultOidcClient. If the name does not match with one of the configured clients, the default client will be autowired.

This means your signature should become:

```php
public function surfconext(Organisation $organisation, OidcClientInterface $adminClient): RedirectResponse
```

**Comment by ajstevens** [comment]
Hi Bob, 

Thank you for the swift response. 

The problem I have is I don't know the correct client at  development time, I can't use the DI unless I make a route per client? What I'm trying to do is launch different clients from the same controller based on the organization people are trying to access, ,so: I have a slug, that links to an organisation, which tells me which OIDC-client to use.
Getting it from the container would be one solution I guess: 
container->get('drenso.oidc.client.default')
container->get('drenso.oidc.client.rkd')

But that would require making the services public and I am wondering what would be the best way. 

Regards, 
Albert-Jan

**Comment by bobvandevijver** [comment]
You can simply inject all the clients you would possibly need in your controller, right?

```php
public function surfconext(
  Organisation $organisation, 
  OidcClientInterface $rkdClient,
  OidcClientInterface $adminClient,
): RedirectResponse
```

It is also possible to define your own public aliases so you can use the container get method (although I would advise against that antipattern).

Another solution would be to provide a global `ClientManager` from this bundle, which can retrieve a client based on a name (a bit like how it is possible to retrieve different Doctrine entity managers by name). Would you be willing to create a PR for that?

**Comment by ajstevens** [comment]
Yes I agree on making the services public. 

In the long run using multiple injections won't scale for us but will do for now. I will make the code work using multiple injections,  and later look into creating a `clientManager`, thanks again for the quick responses.

**Comment by bobvandevijver** [comment]
I hate open issues, so I just released [v2.3.0](https://github.com/Drenso/symfony-oidc/releases/tag/v2.3.0) which now contains a [`OidcClientLocator`](https://github.com/Drenso/symfony-oidc/blob/master/src/OidcClientLocator.php) that should help you out.

```php
  public function surfconext(OidcClientLocator $clientLocator): RedirectResponse
  {
    return $clientLocator->getClient('your_client_id')->generateAuthorizationRedirect();
  }
```

**Comment by ajstevens** [comment]
Wow, thank you. This really does help me out quite a bit.
